### PR TITLE
Add pre-commit config to provide feedback on code quality before pushing

### PR DESCRIPTION
**Pull request by hungdtrn**
In pre-commit config, I use the following hook 
- pre-commit hooks:
  - trailing-whitespace
  - end-of-file-fixer
  - check-byte-order-marker
  - check-added-large-files
  - check-shebang-scripts-are-executable
- ruff hook

### CODE REVIEW

**Comment by nqngo** [comment]
Add Ruff, pre-commit plugins to the VSCode devcontainer.

Add Ruff, pre-commit to dev dependencies.

Specify pre-commit to install on post-install step.

Update from python 3.10 to python 3.12.

Ensure the poetry.lock file is now consistent with pyproject.toml.

**Comment by hungdtrn** [comment]
@nqngo  I have updated devcontainer and pyproject.toml, please have a look.
Also, can you clarify "Specify pre-commit to install on post-install step"? Do you mean that I should write a hook on "post-install" event for pre-commit install?

**Comment by hungdtrn** [comment]
@nqngo I realized that the changes I made had already been done last night

**Comment by nqngo** [comment]
> @nqngo can you clarify "Specify pre-commit to install on post-install step"? Do you mean that I should write a hook on "post-install" event for pre-commit install?

My bad. The comment was around what has been done on my commit and not a task list. Just to clarify what I've setup.

File: poetry.lock [hunk 1]
```diff
@@ -1,9 +1,10 @@
-# This file is automatically @generated by Poetry 1.7.1 and should not be changed by hand.
+# This file is automatically @generated by Poetry 1.4.2 and should not be changed by hand.
```
Comment (samhwang) [inline poetry.lock:1]: We should all be running on Poetry 1.7.1

File: .devcontainer/post-install.sh [hunk 2]
```diff
@@ -1,3 +1,4 @@
 #!/bin/bash
 poetry install
 poetry shell
+pre-commit install
```
Comment (samhwang) [inline .devcontainer/post-install.sh:4]: add an extra line at the end for consistency

**Review by samhwang — CHANGES_REQUESTED** [review]
Add a few scripts in the poetry config file to run ruff, so we can also run it manually.

https://python-poetry.org/docs/cli/#run

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+repos:
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v4.4.0
+  hooks:
+    - id: trailing-whitespace
```
Comment (samhwang) [inline .pre-commit-config.yaml:5]: We shouldn't need to run styling lints through pre-commit-hooks, that should be handled by `ruff`.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    - id: end-of-file-fixer
```
Comment (samhwang) [inline .pre-commit-config.yaml:6]: We shouldn't need to run styling lints through pre-commit-hooks, that should be handled by `ruff`.

File: .devcontainer/post-install.sh (hunk 2 above)
Comment (samhwang) [inline .devcontainer/post-install.sh:4]: Still needs to be updated here

```suggestion
poetry run pre-commit install
```

File: .devcontainer/post-install.sh [hunk 5]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 poetry install
 poetry shell
-pre-commit install
+poetry run pre-commit install
```
Comment (samhwang) [inline .devcontainer/post-install.sh:4]: Missing the extra line at the end of the file

**Comment by hungdtrn** [comment]
> Add a few scripts in the poetry config file to run ruff, so we can also run it manually.
> 
> https://python-poetry.org/docs/cli/#run

Do you have any example of the custom script? I don't know why do we need to have some custom script for running ruff.

**Comment by samhwang** [comment]
> Do you have any example of the custom script? I don't know why do we need to have some custom script for running ruff.

Ignore my previous comment 😓 I had a look at poetry from the `npm` world and thought we can define custom scripts so you don't have to type in `poetry run ruff check .` repeatedly.

File: pyproject.toml [hunk 6]
```diff
@@ -2,16 +2,81 @@
...
+    "buck-out",
+    "build",
+    "dist",
+    "node_modules",
+    "site-packages",
+    "venv",
+]
+
+# Same as Black.
+line-length = 88
```
Comment (samhwang) [inline pyproject.toml:51]: Our current Flake8 runs with 127 line length.

https://github.com/bifrostlab/llm-assistant/blob/334adcaf3abffe9f52ec4927b27df2881b2b0147/.github/workflows/python-app.yml#L34

File: pyproject.toml [hunk 7]
```diff
@@ -2,16 +2,81 @@
...
+quote-style = "double"
+
+# Like Black, indent with spaces, rather than tabs.
+indent-style = "space"
+
+# Like Black, respect magic trailing commas.
+skip-magic-trailing-comma = false
+
+# Like Black, automatically detect the appropriate line ending.
+line-ending = "auto"
```
Comment (nqngo) [inline pyproject.toml:79]: set the line-ending here as "lf" to be consistent with our editor Config.

**Review by samhwang — APPROVED** [review]

**Pull request merged**